### DISCUSSION TITLE
How to get pos and vel, visualization.

### DISCUSSION THREAD

**Discussion by Creasento** [raisimTech/raisimLib#150] [Q&A] [answered]
first, thx to share this amazing simulation! it's so fast and comfortble!

Q :
I am change cpp file and rewrite my own manipulator. I try to use PD control and look Joint's pos and velocity etc. But I don't know how I use getPose or other API. Also it can visualize in unity?(such as graph) Also it can save txt file of pos or vel?

If it can, please tell me how can do it. If you may, please share some examples. :)

P.S I share my PD control cpp code.

```
#include` "raisim/RaisimServer.hpp"
#include "raisim/World.hpp"
#if WIN32
#include <timeapi.h>
#endif

int main(int argc, char* argv[]) {
  auto binaryPath = raisim::Path::setFromArgv(argv[0]);
  raisim::World::setActivationKey(binaryPath.getDirectory() + "\\rsc\\activation.raisim");
#if WIN32
    timeBeginPeriod(1); // for sleep_for function. windows default clock speed is 1/64 second. This sets it to 1ms.
#endif
  
double dt = 0.003;

  raisim::World world; 
  world.setTimeStep(dt);
  world.setERP(world.getTimeStep(), world.getTimeStep());
  world.addGround(0, "steel");
  
  auto arm1 = world.addArticulatedSystem(binaryPath.getDirectory() + "\\rsc\\6dof2\\6dof2.urdf");
  auto arm2 = world.addArticulatedSystem(binaryPath.getDirectory() + "\\rsc\\iiwa\\iiwa14.urdf");

  Eigen::VectorXd jointNominalConfig(5), jointVelocityTarget(arm1->getDOF());
  jointNominalConfig << 0, 0, 0.54, 1.0, 0.0;
  jointVelocityTarget.setZero();

  Eigen::VectorXd jointPgain(arm1->getDOF()), jointDgain(arm1->getDOF());
  jointPgain.tail(5).setConstant(100.0);
  jointDgain.tail(5).setConstant(5.0);

  arm1->setGeneralizedCoordinate(jointNominalConfig);
  arm1->setGeneralizedForce(Eigen::VectorXd::Zero(arm1->getDOF()));
  arm1->setPdGains(jointPgain, jointDgain);
  arm1->setPdTarget(jointNominalConfig, jointVelocityTarget);

  arm1->setName("arm_1");
  arm2->setName("arm_2");
  
  raisim::RaisimServer server(&world);
  server.launchServer();

  world.setMaterialPairProp("steel", "steel", 0.8, 0.95, 0.001);

  for (int i = 0; i < 2000000000000000; i++) {
      std::this_thread::sleep_for(std::chrono::microseconds(1000));
      server.integrateWorldThreadSafe();
  }
  server.killServer();
```

**Comment by jhwangbo** [comment, marked as answer]
All joint frames in the URDF file are stored as ``raisim::CoordinateFrame``. You can call ``getFramePosition``, ``getFrameVelocity``, ``getFrameAngularVelocity`` and so on to get the properties associated with the frame. This should help: https://raisim.com/sections/ArticulatedSystem.html#frames
You can export the whole wolrd. You can find an example here: https://github.com/raisimTech/raisimLib/blob/985bf1f3e3266833db36bf5ab18363edba7742a5/examples/src/server/newtonsCradle.cpp#L117
For visualization, you have to use RaiSimUnity API: https://raisim.com/sections/RaisimServer.html

**Reply by Creasento** [reply]
Thanks for help! I try asap:) I share code if I success.

**Reply by Creasento** [reply]
I added to my code above, but I got an error from world.getFrameIdxByName .
error said
"class::raisim does not have a member getFrameIdxByName"

And below it arm1.getFramePosition(footFrameIndex, footPosition); The part also throws an error.
error said
It says "expression must have a class type, but has type "raisim::ArticulatedSystem *"".

How can I solve this?
I share my code

```
auto footFrameIndex = world.getFrameIdxByName("joint1");

  raisim::Vec<3> footPosition, footVelocity, footAngularVelocity;
  raisim::Mat<3, 3> footOrientation;
  arm1.getFramePosition(footFrameIndex, footPosition);
  arm1.getFrameOrientation(footFrameIndex, footOrientation);
  arm1.getFrameVelocity(footFrameIndex, footVelocity);
  arm1.getFrameAngularVelocity(footFrameIndex, footAngularVelocity);
```

add between "auto arm2..." and "Eigen::Vector..." the code above

**Reply by jhwangbo** [reply]
This sounds like a C++ question. 
1. "class::raisim does not have a member getFrameIdxByName"->because that method is in the articulated system
2. It says "expression must have a class type, but has type "raisim::ArticulatedSystem *"".-> because it is a pointer

**Reply by Creasento** [reply]
I follow the link of https://raisim.com/sections/ArticulatedSystem.html#frames
But I can't solve previous error.
Is there wrong of link example? and how can I use them?

**Reply by jhwangbo** [reply]
I fixed the error on the doc website. Please check it again

**Reply by Creasento** [reply]
Thanks, now move next level. How can I output time-dependent values in a format such as xml or txt?
I try  world.exportToXml(binaryPath.getDirectory(), "exportedWorld.xml"); but it doesn't print what I want.
If I want to print or save some file of Joint vel,pos etc, what API I need to use?

**Reply by jhwangbo** [reply]
There is no such API. You should do it yourself

**Reply by Creasento** [reply]
I try and I think I found some data output code. It print out Powershell page, next step is save some txt or xml files.
I share the code.

enter this code after 

```
 auto footFrameIndex1 = arm1->getFrameIdxByName("joint3");

  raisim::Vec<3> footPosition1, footVelocity1, footAngularVelocity1;
  raisim::Mat<3, 3> footOrientation1;

  arm1->getFrameOrientation(footFrameIndex1, footOrientation1);
  arm1->getFrameVelocity(footFrameIndex1, footVelocity1);
  arm1->getFrameAngularVelocity(footFrameIndex1, footAngularVelocity1);

raisim::RaisimServer server(&world);
  server.launchServer();
```
this code is print joint3 Pos every microsecond 1000 (I think it is right)

```
for (int i = 0; i < 20000; i++) {
      std::this_thread::sleep_for(std::chrono::microseconds(1000));
      arm1->getFramePosition(footFrameIndex1, footPosition1);
      std::cout << footPosition1.e() << std::endl; //cout the position(xyz) of joint3
      //arm1->setGeneralizedVelocity(jointVel1);
      
      server.integrateWorldThreadSafe();
  }
```
